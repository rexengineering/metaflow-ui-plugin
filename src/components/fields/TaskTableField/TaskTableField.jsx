import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import {
  TableContainer,
  Table,
  TableHead,
  TableBody,
  TableRow,
  TableCell,
  Paper,
} from "@material-ui/core";

export const DATA_ERROR_MESSAGE = "There was a problem reading the table data.";

function parseTableData(data) {
  try {
    return JSON.parse(data);
  } catch (error) {
    return { body: [[DATA_ERROR_MESSAGE]] };
  }
}

function TaskTableField({ data }) {
  const [info, setInfo] = useState(parseTableData(data));

  useEffect(() => {
    setInfo(parseTableData(data));
  }, [data]);

  const heading = info?.heading;
  const body = info?.body;

  return (
    <TableContainer component={Paper}>
      <Table>
        {Array.isArray(heading) && (
          <TableHead>
            <TableRow data-testid="table-heading">
              {heading.map((item) => (
                <TableCell key={item}>{item}</TableCell>
              ))}
            </TableRow>
          </TableHead>
        )}
        {Array.isArray(body) && (
          <TableBody>
            {body.map(
              (row) =>
                Array.isArray(row) && (
                  <TableRow key={JSON.stringify(row)} data-testid="table-row">
                    {row.map((item) => (
                      <TableCell key={item} data-testid="table-cell">
                        {item}
                      </TableCell>
                    ))}
                  </TableRow>
                )
            )}
          </TableBody>
        )}
      </Table>
    </TableContainer>
  );
}

TaskTableField.propTypes = {
  data: PropTypes.string.isRequired,
};

export default TaskTableField;
