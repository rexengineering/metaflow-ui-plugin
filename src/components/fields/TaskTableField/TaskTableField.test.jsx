import { render, screen } from "@testing-library/react";
import React from "react";
import TaskTableField from ".";
import { DATA_ERROR_MESSAGE } from "./TaskTableField";

const validJson = JSON.stringify({
  heading: ["item1", "item2", "item3"],
  body: [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
  ],
});

describe("<TaskTextField />", () => {
  it("should render", () => {
    const { asFragment } = render(<TaskTableField />);
    const initialRender = asFragment();

    expect(initialRender).not.toBeNull();
  });

  it("should match snapshot", () => {
    const { asFragment } = render(<TaskTableField data={validJson} />);
    const initialRender = asFragment();

    expect(initialRender).toMatchSnapshot();
  });

  it("should render a table for valid JSON", () => {
    render(<TaskTableField data={validJson} />);

    const rows = screen.queryAllByTestId("table-row");
    rows.forEach((row, rowIndex) => {
      const cells = row.querySelectorAll('[data-testid="table-row"]');
      cells.forEach((cell, cellIndex) => {
        expect(cell).toHaveTextContent(validJson.body[rowIndex][cellIndex]);
      });
    });
  });

  it("should render an error message for invalid JSON", () => {
    render(<TaskTableField data="invalid json" />);

    const row = screen.queryByTestId("table-row");
    expect(row).toHaveTextContent(DATA_ERROR_MESSAGE);
  });

  it("should render a heading if supplied in the JSON", () => {
    render(<TaskTableField data={validJson} />);

    const heading = screen.queryByTestId("table-heading");
    const cells = heading.querySelectorAll('[data-testid="table-row"]');
    cells.forEach((cell, cellIndex) => {
      expect(cell).toHaveTextContent(validJson.heading[cellIndex]);
    });
  });
});
