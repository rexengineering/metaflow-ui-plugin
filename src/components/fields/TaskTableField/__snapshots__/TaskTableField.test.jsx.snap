// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskTextField /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <table
      class="MuiTable-root"
    >
      <thead
        class="MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head"
          data-testid="table-heading"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            scope="col"
          >
            item1
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head"
            scope="col"
          >
            item2
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head"
            scope="col"
          >
            item3
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root"
          data-testid="table-row"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body"
            data-testid="table-cell"
          >
            1
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body"
            data-testid="table-cell"
          >
            2
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body"
            data-testid="table-cell"
          >
            3
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
          data-testid="table-row"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body"
            data-testid="table-cell"
          >
            4
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body"
            data-testid="table-cell"
          >
            5
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body"
            data-testid="table-cell"
          >
            6
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
          data-testid="table-row"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body"
            data-testid="table-cell"
          >
            7
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body"
            data-testid="table-cell"
          >
            8
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body"
            data-testid="table-cell"
          >
            9
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
